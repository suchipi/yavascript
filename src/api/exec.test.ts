import {
  evaluate,
  binaryPath,
  inspect,
  cleanResult,
  TMP,
} from "../test-helpers";

test("exec true - string", async () => {
  const result = await evaluate(`exec("true")`);
  expect(result).toMatchSnapshot();
});

test("exec true - array", async () => {
  const result = await evaluate(`exec(["true"])`);
  expect(result).toMatchSnapshot();
});

test("exec false - string", async () => {
  const result = await evaluate(`exec("false")`);
  expect(result).toMatchSnapshot();
});

test("exec false - array", async () => {
  const result = await evaluate(`exec(["false"])`);
  expect(result).toMatchSnapshot();
});

test("exec - child process receives args", async () => {
  const result = await evaluate(
    `exec([${JSON.stringify(
      binaryPath
    )}, "-e", "scriptArgs", "bla", "blah", "--", "haha"])`
  );
  expect(result).toMatchSnapshot();
});

test("exec with env vars", async () => {
  const result = await evaluate(
    `exec([${JSON.stringify(
      binaryPath
    )}, "-e", "env"], { env: { HI_MOM: "yup" } })`
  );
  expect(result).toMatchSnapshot();
});

test("exec with cwd", async () => {
  const result = await evaluate(
    `exec([${JSON.stringify(binaryPath)}, "-e", "pwd()"], { cwd: "/tmp" })`
  );
  expect(result).toMatchSnapshot();
});

test("exec with env", async () => {
  const result = await evaluate(`exec(['env'], { env: { HI: 'yeah' } })`);
  expect(result).toMatchSnapshot();
});

test("exec with captureOutput true", async () => {
  const result = await evaluate(
    `exec(["echo", "hi there"], { captureOutput: true })`
  );
  expect(result).toMatchSnapshot();
});

test("exec with failOnNonZeroStatus false - running 'false'", async () => {
  const result = await evaluate(
    `exec(["false"], { failOnNonZeroStatus: false })`
  );
  expect(result).toMatchSnapshot();
});

test("exec with failOnNonZeroStatus false - running 'true'", async () => {
  const result = await evaluate(
    `exec(["true"], { failOnNonZeroStatus: false })`
  );
  expect(result).toMatchSnapshot();
});

test("$ echo hi - string", async () => {
  const result = await evaluate(`$(\`echo hi\`)`);
  expect(result).toMatchSnapshot();
});

test("$ echo hi - array", async () => {
  const result = await evaluate(`$(["echo", "hi"])`);
  expect(result).toMatchSnapshot();
});

test("$ false", async () => {
  const result = await evaluate(`$("false")`);
  expect(result).toMatchSnapshot();
});

test("exec parses arg string properly", async () => {
  const result = await evaluate(`exec("bash -c 'echo hi'")`);
  expect(result).toMatchSnapshot();
});

test("exec's string parsing does not interpolate env vars", async () => {
  const result = await evaluate(`exec('echo $HI', { env: { HI: 'yeah' } })`);
  expect(result).toMatchSnapshot();
});

test("exec's string parsing does not parse globs", async () => {
  const result = await evaluate(`exec('echo **/*')`);
  expect(result).toMatchSnapshot();
});

test("logging", async () => {
  const result = await evaluate(
    `exec('echo hi', { trace: console.error }); exec(['echo', '   hi'], { trace: console.error });`
  );
  expect(result).toMatchSnapshot();
});
