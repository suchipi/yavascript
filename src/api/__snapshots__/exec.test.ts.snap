// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$ echo hi - array 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  stdout: \\"hi\\\\n\\"
  stderr: \\"\\"
  status: 0
  signal: undefined
}
",
}
`;

exports[`$ echo hi - string 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  stdout: \\"hi\\\\n\\"
  stderr: \\"\\"
  status: 0
  signal: undefined
}
",
}
`;

exports[`$ false 1`] = `
Object {
  "code": 1,
  "error": false,
  "stderr": "Error: Command failed: [\\"false\\"] (status = 1, signal = undefined)
  at makeErrorWithProperties (<rootDir>/bin/.../yavascript)
  at exec (<rootDir>/bin/.../yavascript)
  at $ (<rootDir>/bin/.../yavascript)
  at <eval> (<evalScript>) {
  status: 1
  signal: undefined
}
",
  "stdout": "",
}
`;

exports[`exec - child process receives args 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "[
  \\"<rootDir>/bin/.../yavascript\\"
  \\"-e\\"
  \\"scriptArgs\\"
  \\"bla\\"
  \\"blah\\"
  \\"--\\"
  \\"haha\\"
]
",
}
`;

exports[`exec false - array 1`] = `
Object {
  "code": 1,
  "error": false,
  "stderr": "Error: Command failed: [\\"false\\"] (status = 1, signal = undefined)
  at makeErrorWithProperties (<rootDir>/bin/.../yavascript)
  at exec (<rootDir>/bin/.../yavascript)
  at <eval> (<evalScript>) {
  status: 1
  signal: undefined
}
",
  "stdout": "",
}
`;

exports[`exec false - string 1`] = `
Object {
  "code": 1,
  "error": false,
  "stderr": "Error: Command failed: [\\"false\\"] (status = 1, signal = undefined)
  at makeErrorWithProperties (<rootDir>/bin/.../yavascript)
  at exec (<rootDir>/bin/.../yavascript)
  at <eval> (<evalScript>) {
  status: 1
  signal: undefined
}
",
  "stdout": "",
}
`;

exports[`exec parses arg string properly 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "hi
",
}
`;

exports[`exec true - array 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`exec true - string 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "",
}
`;

exports[`exec with captureOutput true 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  stdout: \\"hi there\\\\n\\"
  stderr: \\"\\"
  status: 0
  signal: undefined
}
",
}
`;

exports[`exec with cwd 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "/tmp
",
}
`;

exports[`exec with env 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "HI=yeah
",
}
`;

exports[`exec with env vars 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  HI_MOM: \\"yup\\"
}
",
}
`;

exports[`exec with failOnNonZeroStatus false - running 'false' 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  status: 1
  signal: undefined
}
",
}
`;

exports[`exec with failOnNonZeroStatus false - running 'true' 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "{
  status: 0
  signal: undefined
}
",
}
`;

exports[`exec's string parsing does not interpolate env vars 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "$HI
",
}
`;

exports[`exec's string parsing does not parse globs 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "",
  "stdout": "**/*
",
}
`;

exports[`logging 1`] = `
Object {
  "code": 0,
  "error": false,
  "stderr": "ChildProcess.start: [
  \\"echo\\"
  \\"hi\\"
]
ChildProcess result: [
  \\"echo\\"
  \\"hi\\"
] -> {
  status: 0
  signal: undefined
}
ChildProcess.start: [
  \\"echo\\"
  \\"   hi\\"
]
ChildProcess result: [
  \\"echo\\"
  \\"   hi\\"
] -> {
  status: 0
  signal: undefined
}
",
  "stdout": "hi
   hi
",
}
`;
